Argument Clinic: DSLParser.parse_converter() can return an incorrect kwargs dict

_Originally posted by @AlexWaygood in https://github.com/python/cpython/issues/106354#issuecomment-1617797726_:

```
Tools/clinic/clinic.py:5039: error: Incompatible return value type (got
"tuple[str, bool, dict[str | None, Any]]", expected
"tuple[str, bool, dict[str, Any]]")  [return-value]
                    return name, False, kwargs
                           ^~~~~~~~~~~~~~~~~~~
```

I think mypy is flagging a real bug in the code here. The issue is this block of code here:

https://github.com/python/cpython/blob/d694f043ae25308dda607fad4ea1d8d85d79982f/Tools/clinic/clinic.py#L5033-L5039

The function is annotated as return `tuple[str, bool, KwargDict]`, and `KwargDict` is a type alias for `dict[str, Any]`. It's important that the dictionary that's the third element of the tuple only has strings as keys. If it doesn't, then this will fail:

https://github.com/python/cpython/blob/d694f043ae25308dda607fad4ea1d8d85d79982f/Tools/clinic/clinic.py#L4617

The code as written, however, doesn't guarantee that all the keys in the `kwargs` dictionary will be strings. In the dictionary comprehension, we can see that `annotation` is an `ast.Call` instance. That means that `annotation.keywords` is of type `list[ast.keyword]` -- we can see this from typeshed's stubs for the `ast` module (which are an invaluable reference if you're working with ASTs in Python!): https://github.com/python/typeshed/blob/18d45d62aabe68fce78965c4920cbdeddb4b54db/stdlib/_ast.pyi#L324-L329. If `annotation.keywords` is of type `list[ast.keyword]`, that means that the `node` variable in the dictionary comprehension is of type `keyword`, which means (again using typeshed's `ast` stubs), that `node.arg` is of type `str | None`: https://github.com/python/typeshed/blob/18d45d62aabe68fce78965c4920cbdeddb4b54db/stdlib/_ast.pyi#L516-L520. AKA, the keys in this dictionary are not always guaranteed to be strings -- there's a latent bug in this code!

<!-- gh-linked-prs -->
### Linked PRs
* gh-106361
* gh-106364
<!-- /gh-linked-prs -->

--- a/Lib/test/test_clinic.py
+++ b/Lib/test/test_clinic.py
@@ -815,8 +815,8 @@ Annotations must be either a name, a function call, or a string.
 
     def test_kwarg_splats_disallowed_in_function_call_annotations(self):
         expected_error_msg = (
-            "Error on line 0\n"
-            "Cannot use a kwarg splat in a function-call annotation"
+            "Error on line 0:\n"
+            "Cannot use a kwarg splat in a function-call annotation\n"
         )
         dataset = (
             'module fo\nfo.barbaz\n   o: bool(**{None: "bang!"})',

--- a/Lib/test/test_clinic.py
+++ b/Lib/test/test_clinic.py
@@ -797,21 +797,18 @@ Annotations must be either a name, a function call, or a string.
 Error on line 0:
 Annotations must be either a name, a function call, or a string.
 """
-
-        s = self.parse_function_should_fail(
-            'module os\nos.access\n   path: {"some": "dictionary"}'
+        dataset = (
+            'module os\nos.access\n   path: {"some": "dictionary"}',
+            'module os\nos.access\n   path: ["list", "of", "strings"]',
+            'module os\nos.access\n   path: (x for x in range(42))',
+            'module fo\nfo.barbaz\n   o: bool(**{None: "bang!"})',
+            'module fo\nfo.barbaz -> bool(**{None: "bang!"})',
         )
-        self.assertEqual(s, expected_failure_message)
 
-        s = self.parse_function_should_fail(
-            'module os\nos.access\n   path: ["list", "of", "strings"]'
-        )
-        self.assertEqual(s, expected_failure_message)
-
-        s = self.parse_function_should_fail(
-            'module os\nos.access\n   path: (x for x in range(42))'
-        )
-        self.assertEqual(s, expected_failure_message)
+        for fn in dataset:
+            with self.subTest(fn=fn):
+                actual = self.parse_function_should_fail(fn)
+                self.assertEqual(actual, expected_failure_message)
 
     def test_unused_param(self):
         block = self.parse("""

--- a/Lib/test/test_clinic.py
+++ b/Lib/test/test_clinic.py
@@ -821,8 +821,8 @@ Annotations must be either a name, a function call, or a string.
         dataset = (
             'module fo\nfo.barbaz\n   o: bool(**{None: "bang!"})',
             'module fo\nfo.barbaz -> bool(**{None: "bang!"})',
-            'module fo\nfo.barbaz -> bool(**{bang: 42})',
-            'module fo\nfo.barbaz\n   o: bool(**{bang: None})',
+            'module fo\nfo.barbaz -> bool(**{"bang": 42})',
+            'module fo\nfo.barbaz\n   o: bool(**{"bang": None})',
         )
         for fn in dataset:
             with self.subTest(fn=fn):

--- a/Lib/test/test_clinic.py
+++ b/Lib/test/test_clinic.py
@@ -814,7 +814,10 @@ Annotations must be either a name, a function call, or a string.
         self.assertEqual(s, expected_failure_message)
 
     def test_bizarre_parseable_annotations(self):
-        msg = "Cannot use a kwarg splat in a function-call annotation"
+        expected_error_msg = (
+            "Error on line 0\n"
+            "Cannot use a kwarg splat in a function-call annotation"
+        )
         dataset = (
             'module fo\nfo.barbaz\n   o: bool(**{None: "bang!"})',
             'module fo\nfo.barbaz -> bool(**{None: "bang!"})',
@@ -822,8 +825,7 @@ Annotations must be either a name, a function call, or a string.
         for fn in dataset:
             with self.subTest(fn=fn):
                 out = self.parse_function_should_fail(fn)
-                self.assertTrue(out.startswith("Error on line 0"))
-                self.assertIn(msg, out)
+                self.assertEqual(out, expected_error_msg)
 
     def test_unused_param(self):
         block = self.parse("""

--- a/Lib/test/test_clinic.py
+++ b/Lib/test/test_clinic.py
@@ -821,7 +821,7 @@ Annotations must be either a name, a function call, or a string.
         dataset = (
             'module fo\nfo.barbaz\n   o: bool(**{None: "bang!"})',
             'module fo\nfo.barbaz -> bool(**{None: "bang!"})',
-            'module fo\nfo.barbaz -> bool(**{bang: None})',
+            'module fo\nfo.barbaz -> bool(**{bang: 42})',
             'module fo\nfo.barbaz\n   o: bool(**{bang: None})',
         )
         for fn in dataset:

--- a/Lib/test/test_clinic.py
+++ b/Lib/test/test_clinic.py
@@ -821,6 +821,8 @@ Annotations must be either a name, a function call, or a string.
         dataset = (
             'module fo\nfo.barbaz\n   o: bool(**{None: "bang!"})',
             'module fo\nfo.barbaz -> bool(**{None: "bang!"})',
+            'module fo\nfo.barbaz -> bool(**{bang: None})',
+            'module fo\nfo.barbaz\n   o: bool(**{bang: None})',
         )
         for fn in dataset:
             with self.subTest(fn=fn):

--- a/Lib/test/test_clinic.py
+++ b/Lib/test/test_clinic.py
@@ -814,7 +814,7 @@ Annotations must be either a name, a function call, or a string.
         self.assertEqual(s, expected_failure_message)
 
     def test_bizarre_parseable_annotations(self):
-        msg = "Annotation dicts must have str keys"
+        msg = "Cannot use a kwarg splat in a function-call annotation"
         dataset = (
             'module fo\nfo.barbaz\n   o: bool(**{None: "bang!"})',
             'module fo\nfo.barbaz -> bool(**{None: "bang!"})',

--- a/Lib/test/test_clinic.py
+++ b/Lib/test/test_clinic.py
@@ -797,18 +797,33 @@ Annotations must be either a name, a function call, or a string.
 Error on line 0:
 Annotations must be either a name, a function call, or a string.
 """
+
+        s = self.parse_function_should_fail(
+            'module os\nos.access\n   path: {"some": "dictionary"}'
+        )
+        self.assertEqual(s, expected_failure_message)
+
+        s = self.parse_function_should_fail(
+            'module os\nos.access\n   path: ["list", "of", "strings"]'
+        )
+        self.assertEqual(s, expected_failure_message)
+
+        s = self.parse_function_should_fail(
+            'module os\nos.access\n   path: (x for x in range(42))'
+        )
+        self.assertEqual(s, expected_failure_message)
+
+    def test_bizarre_parseable_annotations(self):
+        msg = "Annotation dicts must have str keys"
         dataset = (
-            'module os\nos.access\n   path: {"some": "dictionary"}',
-            'module os\nos.access\n   path: ["list", "of", "strings"]',
-            'module os\nos.access\n   path: (x for x in range(42))',
             'module fo\nfo.barbaz\n   o: bool(**{None: "bang!"})',
             'module fo\nfo.barbaz -> bool(**{None: "bang!"})',
         )
-
         for fn in dataset:
             with self.subTest(fn=fn):
-                actual = self.parse_function_should_fail(fn)
-                self.assertEqual(actual, expected_failure_message)
+                out = self.parse_function_should_fail(fn)
+                self.assertTrue(out.startswith("Error on line 0"))
+                self.assertIn(msg, out)
 
     def test_unused_param(self):
         block = self.parse("""

--- a/Lib/test/test_clinic.py
+++ b/Lib/test/test_clinic.py
@@ -813,7 +813,7 @@ Annotations must be either a name, a function call, or a string.
         )
         self.assertEqual(s, expected_failure_message)
 
-    def test_bizarre_parseable_annotations(self):
+    def test_kwarg_splats_disallowed_in_function_call_annotations(self):
         expected_error_msg = (
             "Error on line 0\n"
             "Cannot use a kwarg splat in a function-call annotation"

--- a/Misc/NEWS.d/next/Tools-Demos/2023-07-03-14-06-19.gh-issue-106359.RfJuR0.rst
+++ b/Misc/NEWS.d/next/Tools-Demos/2023-07-03-14-06-19.gh-issue-106359.RfJuR0.rst
@@ -1,1 +1,2 @@
-Fix bizarre bugs.
+Argument Clinic now explicitly require that keys are string if the provided
+parameter annotation is a dict.

--- a/Misc/NEWS.d/next/Tools-Demos/2023-07-03-14-06-19.gh-issue-106359.RfJuR0.rst
+++ b/Misc/NEWS.d/next/Tools-Demos/2023-07-03-14-06-19.gh-issue-106359.RfJuR0.rst
@@ -1,2 +1,2 @@
-Argument Clinic now explicitly forbids "kwarg splats" in function calls used as 
+Argument Clinic now explicitly forbids "kwarg splats" in function calls used as
 annotations.

--- a/Misc/NEWS.d/next/Tools-Demos/2023-07-03-14-06-19.gh-issue-106359.RfJuR0.rst
+++ b/Misc/NEWS.d/next/Tools-Demos/2023-07-03-14-06-19.gh-issue-106359.RfJuR0.rst
@@ -1,2 +1,2 @@
-Argument Clinic now explicitly require that keys are string if the provided
-parameter annotation is a dict.
+Argument Clinic now explicitly forbids "kwarg splats" in function calls used as 
+annotations

--- a/Misc/NEWS.d/next/Tools-Demos/2023-07-03-14-06-19.gh-issue-106359.RfJuR0.rst
+++ b/Misc/NEWS.d/next/Tools-Demos/2023-07-03-14-06-19.gh-issue-106359.RfJuR0.rst
@@ -1,2 +1,2 @@
 Argument Clinic now explicitly forbids "kwarg splats" in function calls used as 
-annotations
+annotations.

--- a/Misc/NEWS.d/next/Tools-Demos/2023-07-03-14-06-19.gh-issue-106359.RfJuR0.rst
+++ b/Misc/NEWS.d/next/Tools-Demos/2023-07-03-14-06-19.gh-issue-106359.RfJuR0.rst
@@ -1,0 +1,1 @@
+Fix bizarre bugs.

--- a/Tools/clinic/clinic.py
+++ b/Tools/clinic/clinic.py
@@ -5033,10 +5033,11 @@ class DSLParser:
         key = f"{parameter_name}_as_{c_name}" if c_name else parameter_name
         self.function.parameters[key] = p
 
-    KwargDict = dict[str | None, Any]
+    KwargDict = dict[str, Any]
 
     @staticmethod
     def parse_converter(annotation: ast.expr | None) -> tuple[str, bool, KwargDict]:
+        msg = "Annotations must be either a name, a function call, or a string."
         match annotation:
             case ast.Constant(value=str() as value):
                 return value, True, {}
@@ -5044,15 +5045,15 @@ class DSLParser:
                 return name, False, {}
             case ast.Call(func=ast.Name(name)):
                 symbols = globals()
-                kwargs = {
-                    node.arg: eval_ast_expr(node.value, symbols)
-                    for node in annotation.keywords
-                }
+                kwargs: dict[str, Any] = {}
+                for node in annotation.keywords:
+                    if not isinstance(node.arg, str):
+                        fail(msg)
+                    kwargs[node.arg] = eval_ast_expr(node.value, symbols)
+                    print(kwargs)
                 return name, False, kwargs
             case _:
-                fail(
-                    "Annotations must be either a name, a function call, or a string."
-                )
+                fail(msg)
 
     def parse_special_symbol(self, symbol):
         if symbol == '*':

--- a/Tools/clinic/clinic.py
+++ b/Tools/clinic/clinic.py
@@ -5038,7 +5038,6 @@ class DSLParser:
     def parse_converter(
             annotation: ast.expr | None
     ) -> tuple[str, bool, ConverterArgs]:
-        msg = "Annotations must be either a name, a function call, or a string."
         match annotation:
             case ast.Constant(value=str() as value):
                 return value, True, {}
@@ -5049,11 +5048,13 @@ class DSLParser:
                 kwargs: ConverterArgs = {}
                 for node in annotation.keywords:
                     if not isinstance(node.arg, str):
-                        fail(msg)
+                        fail("Annotation dicts must have str keys")
                     kwargs[node.arg] = eval_ast_expr(node.value, symbols)
                 return name, False, kwargs
             case _:
-                fail(msg)
+                fail(
+                    "Annotations must be either a name, a function call, or a string."
+                )
 
     def parse_special_symbol(self, symbol):
         if symbol == '*':

--- a/Tools/clinic/clinic.py
+++ b/Tools/clinic/clinic.py
@@ -5048,7 +5048,7 @@ class DSLParser:
                 kwargs: ConverterArgs = {}
                 for node in annotation.keywords:
                     if not isinstance(node.arg, str):
-                        fail("Annotation dicts must have str keys")
+                        fail("Cannot use a kwarg splat in a function-call annotation")
                     kwargs[node.arg] = eval_ast_expr(node.value, symbols)
                 return name, False, kwargs
             case _:

--- a/Tools/clinic/clinic.py
+++ b/Tools/clinic/clinic.py
@@ -4291,6 +4291,7 @@ class IndentStack:
 
 
 StateKeeper = Callable[[str | None], None]
+ConverterArgs = dict[str, Any]
 
 class DSLParser:
     function: Function | None
@@ -5033,10 +5034,10 @@ class DSLParser:
         key = f"{parameter_name}_as_{c_name}" if c_name else parameter_name
         self.function.parameters[key] = p
 
-    KwargDict = dict[str, Any]
-
     @staticmethod
-    def parse_converter(annotation: ast.expr | None) -> tuple[str, bool, KwargDict]:
+    def parse_converter(
+            annotation: ast.expr | None
+    ) -> tuple[str, bool, ConverterArgs]:
         msg = "Annotations must be either a name, a function call, or a string."
         match annotation:
             case ast.Constant(value=str() as value):
@@ -5045,7 +5046,7 @@ class DSLParser:
                 return name, False, {}
             case ast.Call(func=ast.Name(name)):
                 symbols = globals()
-                kwargs: dict[str, Any] = {}
+                kwargs: ConverterArgs = {}
                 for node in annotation.keywords:
                     if not isinstance(node.arg, str):
                         fail(msg)

--- a/Tools/clinic/clinic.py
+++ b/Tools/clinic/clinic.py
@@ -5051,7 +5051,6 @@ class DSLParser:
                     if not isinstance(node.arg, str):
                         fail(msg)
                     kwargs[node.arg] = eval_ast_expr(node.value, symbols)
-                    print(kwargs)
                 return name, False, kwargs
             case _:
                 fail(msg)

